packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart

```diff
@@ -0,0 +1,823 @@
+  '''
+# ===============================================================================
+# The contents of this file are automatically generated and it is not recommended
+# to modify this file manually.
+# ===============================================================================
+#
+# In order to prevent unexpected splitting of deferred apps, this golden
+# file records the last generated set of loading units. It only possible
+# to obtain the final configuration of loading units after compilation is
+# complete. This means improperly setup imports can only be detected after
+# compilation.
+#
+# This golden file allows the build tool to detect any changes in the generated
+# loading units. During the next build attempt, loading units in this file are
+# compared against the newly generated loading units to check for any new or
+# removed loading units. In the case where loading units do not match, the build
+# will fail and ask the developer to verify that the `deferred-components`
+# configuration in `pubspec.yaml` is correct. Developers should make any necessary
+# changes to integrate new and changed loading units or remove no longer existing
+# loading units from the configuration. The build command should then be
+# re-run to continue the build process.
+#
+# Sometimes, changes to the generated loading units may be unintentional. If
+# the list of loading units in this golden is not what is expected, the app's
+# deferred imports should be reviewed. Third party plugins and packages may
+# also introduce deferred imports that result in unexpected loading units.
+#
+# This list of loading units are a snapshot of the loading units generated
+# by the latest successful gen_snapshot during the assembly of deferred
+# components opted-in apps.
+loading-units:
+''', flush: false, mode: FileMode.append);
+    final Set<int> usedIds = <int>{};
+    for (final LoadingUnit unit in generatedLoadingUnits) {
+      if (usedIds.contains(unit.id)) {
+        continue;
+      }
+      goldenFile.writeAsStringSync('  - id: ${unit.id}\n', flush: false, mode: FileMode.append);
+      goldenFile.writeAsStringSync('    libraries:\n', flush: false, mode: FileMode.append);
+      for (final String lib in unit.libraries) {
+        goldenFile.writeAsStringSync('      - $lib\n', flush: false, mode: FileMode.append);
+      }
+      usedIds.add(unit.id);
+    }
+  }
+
+  void _clearTempDir() {
+    final Directory dir = env.projectDir.childDirectory('build').childDirectory(kDeferredComponentsTempDirectory);
+    if (dir.existsSync()) {
+      dir.deleteSync(recursive: true);
+    }
+  }
+
+  Future<List<String>> _generateDiff() async {
+    final List<String> diffLines = <String>[];
+    String consumeLog(String line) {
+      diffLines.add(line);
+      // Don't pipe output. We will display the results later.
+      return null;
+    }
+    final List<String> command = <String>[];
+    if (globals.platform.isLinux || globals.platform.isMacOS) {
```

Comment:
Do you need a specific diff  format here, or is there a way to do this in Dart? In general, we need to be really careful about introducing new runtime dependencies.

Reply:
I prefer unidiff so that I can recommend a `patch` command to quickly apply all diffs, but if that is not possible the diff can still be displayed in other formats for developer reference. Overall, this step is not necessary for the function of this feature, but it provides helpful quality of life improvements.

I am not aware of any fully-dart ways to do this. The diff commands are more complex than it seems. For unix, `diff` should always be available. On windows however, `fc` is less powerful and doesn't generate the diffs I want. I am exploring using `git diff` instead. Since flutter requires git, git diff seems to be a relatively safe dependency.

In the worst case, if the commands for some reason cannot be run successfully, the diff step can be skipped without any material impact on the results of this validation.

Reply:
I'm actually going to remove the diff task from this PR. I'll land a more fully developed version of it in the future. Meanwhile I'd like to get the rest of the PR moving along.